{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4a30d883",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from simple_function import *\n",
    "import scanpy as sc\n",
    "import os\n",
    "import gzip\n",
    "from bedfile_reader import ReadBED\n",
    "import pyranges as pr\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f832693e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#chosen_region => signal column\n",
    "#regions_start = list of region starting sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5e4ef373",
   "metadata": {},
   "outputs": [],
   "source": [
    "#loading merged regions\n",
    "h3k27ac_gene_merged = read_feather_hm(\"../processing_data/H3K27ac/H3K27ac_peakcount_merged_gene.feather\")\n",
    "h3k27ac_reg_merged = read_feather_hm(\"../processing_data/H3K27ac/H3K27ac_peakcount_merged_reg.feather\")\n",
    "h3k27ac_non_gene_merged = read_feather_hm(\"../processing_data/H3K27ac/H3K27ac_peakcount_merged_none_gene.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b4755ed8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene: (423, 69558)\n",
      "reg: (423, 22940)\n",
      "non_gene: (423, 12)\n"
     ]
    }
   ],
   "source": [
    "print(f\"gene: {h3k27ac_gene_merged.shape}\")\n",
    "print(f\"reg: {h3k27ac_reg_merged.shape}\")\n",
    "print(f\"non_gene: {h3k27ac_non_gene_merged.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b9743d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_to_BED(\"../regions_BED/H3K27ac_gene_domain_expanding.bed\", h3k27ac_gene_domains)\n",
    "#df_to_BED(\"../regions_BED/H3K27ac_reg_domain_expanding.bed\", h3k27ac_reg_domains)\n",
    "#df_to_BED(\"../regions_BED/H3K27ac_non_gene_domain_expanding.bed\", h3k27ac_non_gene_domains)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a2b14e06",
   "metadata": {},
   "source": [
    "## set up \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8b07d1d7",
   "metadata": {},
   "source": [
    "### get signalvalue by reading files and extract peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dafde383",
   "metadata": {},
   "outputs": [],
   "source": [
    "def average_signalValue(data: pd.DataFrame, bedfiles_path: str) -> pd.DataFrame:\n",
    "    #data: pd.DataFrame with region and peak count from each sample, sorted columns accrodding to their position of chr1\n",
    "    #bedfiles_path: path to the sorted bedfile of sample\n",
    "    \n",
    "    #creating signalValue dataframe\n",
    "    signal_df = pd.DataFrame(0, index = data.index, columns = data.columns)\n",
    "    \n",
    "    # get starting site of each columns\n",
    "    chosen_regions = list(signal_df.columns)\n",
    "    regions = []\n",
    "    start_sites = []\n",
    "    for region in chosen_regions:\n",
    "        _, pos = region.split(':')\n",
    "        s, e = pos.split('-')\n",
    "        start_sites.append(s)\n",
    "        regions.append([int(s), int(e)])\n",
    "\n",
    "    #np.array of start sites\n",
    "    start_sites = np.array(start_sites).astype('int')\n",
    "    \n",
    "    #file list in path\n",
    "    hm_files = os.listdir(bedfiles_path)\n",
    "    \n",
    "    # getting signalValue\n",
    "    # sorted file -> chr1 appears first \n",
    "    for f in hm_files:\n",
    "        name = f.split('_')[1].split('.bed')[0]\n",
    "        signalValue = []\n",
    "        with gzip.open(f'{bedfiles_path}/{f}', 'rb') as peaks:\n",
    "            lines = peaks.readlines()\n",
    "            for line in lines:\n",
    "                line = line.decode('utf-8')\n",
    "                if line.startswith('chr1\\t'):\n",
    "                    start, end, signalValue, peak_pos = [line.split('\\t')[i] for i in [1, 2, 6, 9]]\n",
    "                    peak = int(start) + int(peak_pos)\n",
    "                    where = np.searchsorted(start_sites, peak)\n",
    "                    #print(signalValue, peak, where)\n",
    "                    if where != 0 and peak <= regions[where-1][1]:\n",
    "                        addValue = float(signalValue)\n",
    "                        signal_df.loc[name, chosen_regions[where-1]] = signal_df.loc[name, chosen_regions[where-1]] + addValue\n",
    "                else:\n",
    "                    break #after looping all chr1 peaks then stop\n",
    "    \n",
    "    # calculate average = signalValue/number of peaks\n",
    "    average = signal_df.values / data.values\n",
    "    average = np.nan_to_num(average, 0)\n",
    "    average = pd.DataFrame(average, index = signal_df.index, columns = signal_df.columns)\n",
    "    \n",
    "    return average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbabc2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "avg_signalValue_h3k27ac_gene = average_signalValue(h3k27ac_gene_merged, H3K27ac_sorted_path)\n",
    "#69k columns takes 53 mins"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e3ff8ee0",
   "metadata": {},
   "source": [
    "### using pyranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f8381575",
   "metadata": {},
   "outputs": [],
   "source": [
    "def average_signalValue_pyranges(data: pd.DataFrame, bedfile_path: str) -> pd.DataFrame:\n",
    "    #data: pd.DataFrame with region and peak count from each sample, sorted columns accrodding to their position of chr1\n",
    "    #bedfiles_path: path to the sorted bedfile of sample\n",
    "    count = 1\n",
    "    \n",
    "    #creating signalValue dataframe\n",
    "    #signal_df = data.copy()\n",
    "    signal_list = []\n",
    "       \n",
    "    #file list in path\n",
    "    hm_files = os.listdir(bedfile_path)\n",
    "    \n",
    "    #string represent of region, region df [chrom, start, end]\n",
    "    region_df = list_domain_todf(list(data.columns))\n",
    "    region_df.columns = ['Chromosome', 'Start', 'End']\n",
    "    region_pr = pr.PyRanges(region_df)\n",
    "    region_pr.__setattr__('Name', list(data.columns))\n",
    "    \n",
    "    \n",
    "    # getting signalValue\n",
    "    # sorted file -> chr1 appears first - not necessary  \n",
    "    for file in hm_files:\n",
    "        print(1, end='\\r', flush=True)\n",
    "        #sample id\n",
    "        name = file.split('_')[1].split('.bed')[0]\n",
    "        \n",
    "        #read sample in pyranges. get peak positions table\n",
    "        sample = ReadBED(f\"{bedfile_path}/{file}\")\n",
    "        peaks = sample.get_peak_data()\n",
    "        \n",
    "        print(2, end='\\r', flush=True)\n",
    "        #extract peak overlap with which region: intersect() -> merge with peak table to get annotate region into matching peak\n",
    "        peak_overlap_region = region_pr.intersect(peaks).df\n",
    "        if peak_overlap_region.empty:\n",
    "            continue\n",
    "            \n",
    "        peak_region_match = peaks.df.merge(\n",
    "            peak_overlap_region[['Chromosome', 'Start', 'End', 'Name']], on=['Chromosome', 'Start', 'End'])  \n",
    "        \n",
    "        print(3, end='\\r', flush=True)\n",
    "        #calculate sum of signal value\n",
    "        sum_signal_value = peak_region_match[['Chromosome', 'signalValue', 'Name']].groupby(['Name']).sum()\n",
    "        \n",
    "        print(4, end='\\r', flush=True)\n",
    "        #match signal to regions\n",
    "        signal_to_region = pd.merge(region_pr.df, sum_signal_value, how='left', on=\"Name\")\n",
    "        signal_to_region['signalValue'] = signal_to_region['signalValue'].fillna(value=0)\n",
    "        #print(signal_to_region, flush=True)\n",
    "        \n",
    "        print(5, end='\\r', flush=True)\n",
    "        #replace sum signalValue \n",
    "        #signal_df.loc[name, :] = signal_to_region['signalValue'].to_numpy()\n",
    "        signal = signal_to_region['signalValue'].to_numpy()\n",
    "        avg = signal / data.loc[name, :]\n",
    "        avg = avg.fillna(value=0)\n",
    "        row = [name] + list(avg)\n",
    "        signal_list.append(row)        \n",
    "        \n",
    "        print(6, end='\\r', flush=True)\n",
    "        print(f\"done with {count} files over {len(hm_files)}\", end='\\r')\n",
    "        count += 1 \n",
    "        \n",
    "    print(7, end='\\r', flush=True)  \n",
    "    #create dataframe\n",
    "    fields = ['Sample_ID'] + list(data.columns)\n",
    "    signal_df = pd.DataFrame(signal_list, columns = fields)\n",
    "    signal_df = signal_df.set_index('Sample_ID')\n",
    "  \n",
    "    print(f\"Done processing\", end=\"\\r\")                \n",
    "    return signal_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4a7a4031",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 12s, sys: 19.1 s, total: 4min 31s\n",
      "Wall time: 4min 7s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pr_signalValue_h3k27ac_gene_ = average_signalValue_pyranges(h3k27ac_gene_merged, H3K27ac_sorted_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a86a810f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processinges over 42323\r"
     ]
    }
   ],
   "source": [
    "pr_signalValue_h3k27ac_reg = average_signalValue_pyranges(h3k27ac_reg_merged, H3K27ac_sorted_path)\n",
    "pr_signalValue_h3k27ac_non_gene = average_signalValue_pyranges(h3k27ac_non_gene_merged, H3K27ac_sorted_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f268b15a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done saving dataframe as: ../processing_data/H3K27ac_avgSignalValue_merged_none_gene.feather\n",
      "Done saving dataframe as: ../processing_data/H3K27ac_avgSignalValue_merged_gene.feather\n",
      "Done saving dataframe as: ../processing_data/H3K27ac_avgSignalValue_merged_reg.feather\n"
     ]
    }
   ],
   "source": [
    "#save file\n",
    "write_feather_df(pr_signalValue_h3k27ac_non_gene, '../processing_data/H3K27ac_avgSignalValue_merged_none_gene.feather')\n",
    "write_feather_df(pr_signalValue_h3k27ac_gene_, '../processing_data/H3K27ac_avgSignalValue_merged_gene.feather')\n",
    "write_feather_df(pr_signalValue_h3k27ac_reg, '../processing_data/H3K27ac_avgSignalValue_merged_reg.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51e1a2f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>chr1:123686524-123687523</th>\n",
       "      <th>chr1:123719524-123720523</th>\n",
       "      <th>chr1:123802524-123803523</th>\n",
       "      <th>chr1:123892524-123893523</th>\n",
       "      <th>chr1:124367524-124368523</th>\n",
       "      <th>chr1:124410524-124411523</th>\n",
       "      <th>chr1:124720524-124721523</th>\n",
       "      <th>chr1:125007524-125008523</th>\n",
       "      <th>chr1:125064524-125065523</th>\n",
       "      <th>chr1:125066524-125067523</th>\n",
       "      <th>chr1:125084524-125085523</th>\n",
       "      <th>chr1:125161524-125162523</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENCFF064KHS</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.56989</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF434KGU</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.26821</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF008FVK</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.19806</td>\n",
       "      <td>3.30716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF115DDQ</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.12763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENCFF051OAR</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.28885</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             chr1:123686524-123687523  chr1:123719524-123720523  chr1:123802524-123803523  \\\n",
       "Sample_ID                                                                                   \n",
       "ENCFF064KHS  0.0                       0.00000                   0.0                        \n",
       "ENCFF434KGU  0.0                       0.00000                   0.0                        \n",
       "ENCFF008FVK  0.0                       0.00000                   0.0                        \n",
       "ENCFF115DDQ  0.0                       2.12763                   0.0                        \n",
       "ENCFF051OAR  0.0                       0.00000                   0.0                        \n",
       "\n",
       "             chr1:123892524-123893523  chr1:124367524-124368523  chr1:124410524-124411523  \\\n",
       "Sample_ID                                                                                   \n",
       "ENCFF064KHS  0.0                       0.0                       0.0                        \n",
       "ENCFF434KGU  0.0                       0.0                       0.0                        \n",
       "ENCFF008FVK  0.0                       0.0                       0.0                        \n",
       "ENCFF115DDQ  0.0                       0.0                       0.0                        \n",
       "ENCFF051OAR  0.0                       0.0                       0.0                        \n",
       "\n",
       "             chr1:124720524-124721523  chr1:125007524-125008523  chr1:125064524-125065523  \\\n",
       "Sample_ID                                                                                   \n",
       "ENCFF064KHS  0.00000                   0.00000                   3.56989                    \n",
       "ENCFF434KGU  0.00000                   4.26821                   0.00000                    \n",
       "ENCFF008FVK  0.00000                   0.00000                   0.00000                    \n",
       "ENCFF115DDQ  0.00000                   0.00000                   0.00000                    \n",
       "ENCFF051OAR  7.28885                   0.00000                   0.00000                    \n",
       "\n",
       "             chr1:125066524-125067523  chr1:125084524-125085523  chr1:125161524-125162523  \n",
       "Sample_ID                                                                                  \n",
       "ENCFF064KHS  0.0                       0.00000                   0.00000                   \n",
       "ENCFF434KGU  0.0                       0.00000                   0.00000                   \n",
       "ENCFF008FVK  0.0                       4.19806                   3.30716                   \n",
       "ENCFF115DDQ  0.0                       0.00000                   0.00000                   \n",
       "ENCFF051OAR  0.0                       0.00000                   0.00000                   "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pr_signalValue_h3k27ac_non_gene.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5d54a0c8",
   "metadata": {},
   "source": [
    "-------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2922b130",
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k27me3_sorted_path = \"../data/sorted_H3K27me3/\"\n",
    "region_type = ['gene', 'reg', 'none_gene']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6ecfa7a5",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "h3k27me3_gene_merged = read_feather_hm(\"../processing_data/H3K27me3/H3K27me3_peakcount_merged_gene.feather\")\n",
    "h3k27me3_reg_merged = read_feather_hm(\"../processing_data/H3K27me3/H3K27me3_peakcount_merged_reg.feather\")\n",
    "h3k27me3_non_gene_merged = read_feather_hm(\"../processing_data/H3K27me3/H3K27me3_peakcount_merged_none_gene.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a010c069",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done saving dataframe as: ../processing_data/H3K27me3/H3K27me3_avgSignalValue_merged_gene.feather\n",
      "Done saving dataframe as: ../processing_data/H3K27me3/H3K27me3_avgSignalValue_merged_reg.feather\n",
      "Done saving dataframe as: ../processing_data/H3K27me3/H3K27me3_avgSignalValue_merged_none_gene.feather\n"
     ]
    }
   ],
   "source": [
    "for t in region_type:\n",
    "    if t == 'gene':\n",
    "        avg_sv = average_signalValue_pyranges(h3k27me3_gene_merged, h3k27me3_sorted_path)\n",
    "    elif t == 'reg':\n",
    "        avg_sv = average_signalValue_pyranges(h3k27me3_reg_merged, h3k27me3_sorted_path)\n",
    "    elif t == 'none_gene':\n",
    "        avg_sv = average_signalValue_pyranges(h3k27me3_non_gene_merged, h3k27me3_sorted_path)\n",
    "    write_feather_df(avg_sv, f'../processing_data/H3K27me3/H3K27me3_avgSignalValue_merged_{t}.feather')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5bb0c081",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Assay name</th>\n",
       "      <th>Assay title</th>\n",
       "      <th>Biosample classification</th>\n",
       "      <th>Target</th>\n",
       "      <th>Target of assay</th>\n",
       "      <th>Target gene symbol</th>\n",
       "      <th>Biosample summary</th>\n",
       "      <th>Biosample term name</th>\n",
       "      <th>Dbxrefs</th>\n",
       "      <th>Project</th>\n",
       "      <th>Files</th>\n",
       "      <th>Biosample accession</th>\n",
       "      <th>Biological replicate</th>\n",
       "      <th>Technical replicate</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Life stage</th>\n",
       "      <th>Replicates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/experiments/ENCSR965BLU/</td>\n",
       "      <td>ENCSR965BLU</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens gastroesophageal sphincter tissue...</td>\n",
       "      <td>gastroesophageal sphincter</td>\n",
       "      <td>GEO:GSE139793</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF862NQF/,/files/ENCFF425WPG/,/files...</td>\n",
       "      <td>ENCBS727XQL</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/2ba173e8-0aab-46a2-8e30-b8e4169162...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/experiments/ENCSR567VAB/</td>\n",
       "      <td>ENCSR567VAB</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>GEO:GSE209056</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF786FYS/,/files/ENCFF759PXY/,/files...</td>\n",
       "      <td>ENCBS464LBN</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/0dccda4d-b275-4695-87b2-fafef646de...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/experiments/ENCSR532EYS/</td>\n",
       "      <td>ENCSR532EYS</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>GEO:GSE208927</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF754HYY/,/files/ENCFF979WLU/,/files...</td>\n",
       "      <td>ENCBS485UEI</td>\n",
       "      <td>1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/4bacc45c-3138-4d06-82fa-5e2361cd68...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/experiments/ENCSR558GCO/</td>\n",
       "      <td>ENCSR558GCO</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens dorsolateral prefrontal cortex ti...</td>\n",
       "      <td>dorsolateral prefrontal cortex</td>\n",
       "      <td>GEO:GSE187693</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF607HEJ/,/files/ENCFF983QXT/,/files...</td>\n",
       "      <td>ENCBS733LER</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/1f1ddc69-6894-4308-8306-0b1a53151e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/experiments/ENCSR469YCE/</td>\n",
       "      <td>ENCSR469YCE</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens upper lobe of left lung tissue ma...</td>\n",
       "      <td>upper lobe of left lung</td>\n",
       "      <td>GEO:GSE139780</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF128ZRT/,/files/ENCFF602SXE/,/files...</td>\n",
       "      <td>ENCBS943JOF</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/8446d46d-884d-4d12-8f1b-93afd9d66e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>/experiments/ENCSR188HXK/</td>\n",
       "      <td>ENCSR188HXK</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens esophagus muscularis mucosa tissu...</td>\n",
       "      <td>esophagus muscularis mucosa</td>\n",
       "      <td>GEO:GSE139724</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF161HMR/,/files/ENCFF535WCS/,/files...</td>\n",
       "      <td>ENCBS884JOD</td>\n",
       "      <td>1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/2551faef-dac7-40e2-831c-b3720f4c35...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>/experiments/ENCSR632SLJ/</td>\n",
       "      <td>ENCSR632SLJ</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens Peyer's patch tissue male adult (...</td>\n",
       "      <td>Peyer's patch</td>\n",
       "      <td>GEO:GSE139711</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF329VRW/,/files/ENCFF268ZAW/,/files...</td>\n",
       "      <td>ENCBS388XLV</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/6bef690b-d6bb-4aea-8329-e4350cc767...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>/experiments/ENCSR853JJZ/</td>\n",
       "      <td>ENCSR853JJZ</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens chorionic villus tissue female em...</td>\n",
       "      <td>chorionic villus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>/files/SRR2172735/,/files/SRR2172736/,/files/E...</td>\n",
       "      <td>ENCBS412DCK</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>/replicates/df5f9e4a-f355-463b-9caf-f6b853847f8f/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>/experiments/ENCSR673XMI/</td>\n",
       "      <td>ENCSR673XMI</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens chorionic villus tissue male embr...</td>\n",
       "      <td>chorionic villus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>/files/SRR2172763/,/files/SRR2172764/,/files/E...</td>\n",
       "      <td>ENCBS706YBS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>/replicates/fee6de29-264d-4455-b3e9-741234f4ea57/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>/experiments/ENCSR123ASL/</td>\n",
       "      <td>ENCSR123ASL</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens spleen tissue female adult (41 ye...</td>\n",
       "      <td>spleen</td>\n",
       "      <td>GEO:GSE167709</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF306ZIE/,/files/ENCFF873KMX/,/files...</td>\n",
       "      <td>ENCBS248EXJ</td>\n",
       "      <td>1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/acb0b8fa-69dc-4f18-b2b5-4a7e6fae0a...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>216 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID                         Accession    Assay name Assay title       Biosample classification  \\\n",
       "0    /experiments/ENCSR965BLU/  ENCSR965BLU  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "1    /experiments/ENCSR567VAB/  ENCSR567VAB  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "2    /experiments/ENCSR532EYS/  ENCSR532EYS  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "3    /experiments/ENCSR558GCO/  ENCSR558GCO  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "4    /experiments/ENCSR469YCE/  ENCSR469YCE  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "..                         ...          ...        ...               ...                      ...   \n",
       "211  /experiments/ENCSR188HXK/  ENCSR188HXK  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "212  /experiments/ENCSR632SLJ/  ENCSR632SLJ  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "213  /experiments/ENCSR853JJZ/  ENCSR853JJZ  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "214  /experiments/ENCSR673XMI/  ENCSR673XMI  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "215  /experiments/ENCSR123ASL/  ENCSR123ASL  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "\n",
       "    Target                    Target of assay Target gene symbol           \\\n",
       "0    /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "1    /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "2    /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "3    /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "4    /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "..                        ...             ...                         ...   \n",
       "211  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "212  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "213  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "214  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "215  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "\n",
       "    Biosample summary                                  Biosample term name              \\\n",
       "0    Homo sapiens gastroesophageal sphincter tissue...      gastroesophageal sphincter   \n",
       "1    Homo sapiens dorsolateral prefrontal cortex ti...  dorsolateral prefrontal cortex   \n",
       "2    Homo sapiens dorsolateral prefrontal cortex ti...  dorsolateral prefrontal cortex   \n",
       "3    Homo sapiens dorsolateral prefrontal cortex ti...  dorsolateral prefrontal cortex   \n",
       "4    Homo sapiens upper lobe of left lung tissue ma...         upper lobe of left lung   \n",
       "..                                                 ...                             ...   \n",
       "211  Homo sapiens esophagus muscularis mucosa tissu...     esophagus muscularis mucosa   \n",
       "212  Homo sapiens Peyer's patch tissue male adult (...                   Peyer's patch   \n",
       "213  Homo sapiens chorionic villus tissue female em...                chorionic villus   \n",
       "214  Homo sapiens chorionic villus tissue male embr...                chorionic villus   \n",
       "215  Homo sapiens spleen tissue female adult (41 ye...                          spleen   \n",
       "\n",
       "    Dbxrefs        Project  Files                                               \\\n",
       "0    GEO:GSE139793   ENCODE  /files/ENCFF862NQF/,/files/ENCFF425WPG/,/files...   \n",
       "1    GEO:GSE209056   ENCODE  /files/ENCFF786FYS/,/files/ENCFF759PXY/,/files...   \n",
       "2    GEO:GSE208927   ENCODE  /files/ENCFF754HYY/,/files/ENCFF979WLU/,/files...   \n",
       "3    GEO:GSE187693   ENCODE  /files/ENCFF607HEJ/,/files/ENCFF983QXT/,/files...   \n",
       "4    GEO:GSE139780   ENCODE  /files/ENCFF128ZRT/,/files/ENCFF602SXE/,/files...   \n",
       "..             ...      ...                                                ...   \n",
       "211  GEO:GSE139724   ENCODE  /files/ENCFF161HMR/,/files/ENCFF535WCS/,/files...   \n",
       "212  GEO:GSE139711   ENCODE  /files/ENCFF329VRW/,/files/ENCFF268ZAW/,/files...   \n",
       "213            NaN  Roadmap  /files/SRR2172735/,/files/SRR2172736/,/files/E...   \n",
       "214            NaN  Roadmap  /files/SRR2172763/,/files/SRR2172764/,/files/E...   \n",
       "215  GEO:GSE167709   ENCODE  /files/ENCFF306ZIE/,/files/ENCFF873KMX/,/files...   \n",
       "\n",
       "    Biosample accession Biological replicate Technical replicate Organism      Life stage  \\\n",
       "0    ENCBS727XQL         1                    1,2                 Homo sapiens      adult   \n",
       "1    ENCBS464LBN         1                    1,2                 Homo sapiens      adult   \n",
       "2    ENCBS485UEI         1                    2,1                 Homo sapiens      adult   \n",
       "3    ENCBS733LER         1                    1,2                 Homo sapiens      adult   \n",
       "4    ENCBS943JOF         1                    1,2                 Homo sapiens      adult   \n",
       "..                  ...                  ...                 ...           ...        ...   \n",
       "211  ENCBS884JOD         1                    2,1                 Homo sapiens      adult   \n",
       "212  ENCBS388XLV         1                    1,2                 Homo sapiens      adult   \n",
       "213  ENCBS412DCK         1                      1                 Homo sapiens  embryonic   \n",
       "214  ENCBS706YBS         1                      1                 Homo sapiens  embryonic   \n",
       "215  ENCBS248EXJ         1                    1,2                 Homo sapiens      adult   \n",
       "\n",
       "    Replicates                                          \n",
       "0    /replicates/2ba173e8-0aab-46a2-8e30-b8e4169162...  \n",
       "1    /replicates/0dccda4d-b275-4695-87b2-fafef646de...  \n",
       "2    /replicates/4bacc45c-3138-4d06-82fa-5e2361cd68...  \n",
       "3    /replicates/1f1ddc69-6894-4308-8306-0b1a53151e...  \n",
       "4    /replicates/8446d46d-884d-4d12-8f1b-93afd9d66e...  \n",
       "..                                                 ...  \n",
       "211  /replicates/2551faef-dac7-40e2-831c-b3720f4c35...  \n",
       "212  /replicates/6bef690b-d6bb-4aea-8329-e4350cc767...  \n",
       "213  /replicates/df5f9e4a-f355-463b-9caf-f6b853847f8f/  \n",
       "214  /replicates/fee6de29-264d-4455-b3e9-741234f4ea57/  \n",
       "215  /replicates/acb0b8fa-69dc-4f18-b2b5-4a7e6fae0a...  \n",
       "\n",
       "[216 rows x 19 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h3k27me3_exp = pd.read_csv('../Histone_data_tsv/experiment_report_H3K27me3.tsv', delimiter='\\t', skiprows=1)\n",
    "h3k27me3_exp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f9ce65fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Assay name</th>\n",
       "      <th>Assay title</th>\n",
       "      <th>Biosample classification</th>\n",
       "      <th>Target</th>\n",
       "      <th>Target of assay</th>\n",
       "      <th>Target gene symbol</th>\n",
       "      <th>Biosample summary</th>\n",
       "      <th>Biosample term name</th>\n",
       "      <th>Dbxrefs</th>\n",
       "      <th>Project</th>\n",
       "      <th>Files</th>\n",
       "      <th>Biosample accession</th>\n",
       "      <th>Biological replicate</th>\n",
       "      <th>Technical replicate</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Life stage</th>\n",
       "      <th>Replicates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>/experiments/ENCSR453MSI/</td>\n",
       "      <td>ENCSR453MSI</td>\n",
       "      <td>ChIP-seq</td>\n",
       "      <td>Histone ChIP-seq</td>\n",
       "      <td>tissue</td>\n",
       "      <td>/targets/H3K27me3-human/</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>HIST2H3A,HIST2H3D,HIST2H3C</td>\n",
       "      <td>Homo sapiens gastrocnemius medialis tissue fem...</td>\n",
       "      <td>gastrocnemius medialis</td>\n",
       "      <td>GEO:GSE142979</td>\n",
       "      <td>ENCODE</td>\n",
       "      <td>/files/ENCFF157ZPT/,/files/ENCFF862IWQ/,/files...</td>\n",
       "      <td>ENCBS640TBC,ENCBS731GTN</td>\n",
       "      <td>2,1</td>\n",
       "      <td>1,2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>adult</td>\n",
       "      <td>/replicates/968a1f1c-3a63-4380-a2f4-8dd3890830...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                         Accession    Assay name Assay title       Biosample classification  \\\n",
       "11  /experiments/ENCSR453MSI/  ENCSR453MSI  ChIP-seq   Histone ChIP-seq  tissue                    \n",
       "\n",
       "   Target                    Target of assay Target gene symbol           \\\n",
       "11  /targets/H3K27me3-human/  H3K27me3        HIST2H3A,HIST2H3D,HIST2H3C   \n",
       "\n",
       "   Biosample summary                                  Biosample term name     Dbxrefs         \\\n",
       "11  Homo sapiens gastrocnemius medialis tissue fem...  gastrocnemius medialis  GEO:GSE142979   \n",
       "\n",
       "   Project Files                                              Biosample accession       \\\n",
       "11  ENCODE  /files/ENCFF157ZPT/,/files/ENCFF862IWQ/,/files...  ENCBS640TBC,ENCBS731GTN   \n",
       "\n",
       "   Biological replicate Technical replicate Organism      Life stage  \\\n",
       "11  2,1                  1,2                 Homo sapiens  adult       \n",
       "\n",
       "   Replicates                                          \n",
       "11  /replicates/968a1f1c-3a63-4380-a2f4-8dd3890830...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h3k27me3_exp.query(f'Files.str.contains(\"ENCFF964DID\")')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "50caa691",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_h3k27me3 = os.listdir('../data/H3K27me3/')\n",
    "infor = {'Code':[],\n",
    "         'Experiment':[],\n",
    "         'Target':[],\n",
    "         'Tissue':[],\n",
    "         'Biosample_summary':[]}\n",
    "for f in file_h3k27me3:\n",
    "    assay_id = f.split('.bed')[0]\n",
    "    retrieve = h3k27me3_exp.query(f'Files.str.contains(\"{assay_id}\")')\n",
    "    infor['Code'].append(assay_id)\n",
    "    infor['Experiment'].append(retrieve['Accession'].item())\n",
    "    infor['Target'].append(retrieve['Target of assay'].item())\n",
    "    infor['Tissue'].append(retrieve['Biosample term name'].item())\n",
    "    infor['Biosample_summary'].append(retrieve['Biosample summary']) \n",
    "\n",
    "h3k27me3_info_df = pd.DataFrame.from_dict(infor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b92367eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Code</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>Target</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Biosample_summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCFF765YZY</td>\n",
       "      <td>ENCSR627STE</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>brain</td>\n",
       "      <td>208    Homo sapiens brain tissue male embryo (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCFF958KOY</td>\n",
       "      <td>ENCSR042RIW</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>sigmoid colon</td>\n",
       "      <td>64    Homo sapiens sigmoid colon tissue male a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCFF148VRO</td>\n",
       "      <td>ENCSR177CGN</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>esophagus muscularis mucosa</td>\n",
       "      <td>142    Homo sapiens esophagus muscularis mucos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCFF608JQT</td>\n",
       "      <td>ENCSR975GDL</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>left lung</td>\n",
       "      <td>179    Homo sapiens left lung tissue male adul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCFF206GAK</td>\n",
       "      <td>ENCSR611CHC</td>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>right lobe of liver</td>\n",
       "      <td>173    Homo sapiens right lobe of liver tissue...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Code         Experiment   Target    Tissue                        \\\n",
       "0  ENCFF765YZY  ENCSR627STE  H3K27me3                        brain   \n",
       "1  ENCFF958KOY  ENCSR042RIW  H3K27me3                sigmoid colon   \n",
       "2  ENCFF148VRO  ENCSR177CGN  H3K27me3  esophagus muscularis mucosa   \n",
       "3  ENCFF608JQT  ENCSR975GDL  H3K27me3                    left lung   \n",
       "4  ENCFF206GAK  ENCSR611CHC  H3K27me3          right lobe of liver   \n",
       "\n",
       "  Biosample_summary                                   \n",
       "0  208    Homo sapiens brain tissue male embryo (...  \n",
       "1  64    Homo sapiens sigmoid colon tissue male a...  \n",
       "2  142    Homo sapiens esophagus muscularis mucos...  \n",
       "3  179    Homo sapiens left lung tissue male adul...  \n",
       "4  173    Homo sapiens right lobe of liver tissue...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h3k27me3_info_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "be089de0",
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k27me3_info_df.to_csv('../Histone_data_tsv/H3K27me3_tissue_info.csv', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a4c751a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
